pkg/envoy/registry/services.go

```diff
@@ -0,0 +1,98 @@
+package registry
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/openservicemesh/osm/pkg/envoy"
+	k8s "github.com/openservicemesh/osm/pkg/kubernetes"
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+// ProxyServiceMapper knows how to map Envoy instances to services.
+type ProxyServiceMapper interface {
+	ListProxyServices(*envoy.Proxy) ([]service.MeshService, error)
+}
+
+// ExplicitProxyServiceMapper is a custom ProxyServiceMapper implementation.
+type ExplicitProxyServiceMapper func(*envoy.Proxy) ([]service.MeshService, error)
+
+// ListProxyServices executes the given mapping.
+func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
+	return e(p)
+}
+
+// KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
+type KubeProxyServiceMapper struct {
+	KubeController k8s.Controller
+}
+
+// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
+func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
```

Comment:
I would imagine `GetServices()` to be implemented o the `envoy.Proxy` object. The `KubeProxyServiceMapper` seems avoidable given it only encapsulates `kubernetes.Controller`.

Reply:
That's not a bad idea. 
So instead of `service, err := proxyRegistry.ListProxyServices(proxy)` we can have `service, err := proxy.GetServices()`?

@nojnhuh if you also agree this is a good simplification - perhaps feel free to open a follow-up GitHub issue to re-refactor this in another PR.

Reply:
> I would imagine `GetServices()` to be implemented o the `envoy.Proxy` object.

That should be simple to change. My main reasoning for doing it this way was that potential related functionality (perhaps `ListServiceProxies()`) makes more sense to me implemented in a layer above a single `Proxy` and maintaining that related behavior in different places seems less than ideal. In this case I can see where `proxy.GetServices()` is more ergonomic though.

> The `KubeProxyServiceMapper` seems avoidable given it only encapsulates `kubernetes.Controller`.

Do you propose this functionality be implemented directly against `kubernetes.Controller`? That would involve exposing Envoy details there, which I mentioned in the other thread.

Reply:
Don't want to complicate this, let's merge this :-)